### ISSUE TITLE
[Phase 7] Production Inference

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/ecg-arrhythmia-classification#8]
## Phase 7: Production Inference

**Parent:** #1
**Depends on:** #7

### Objectives
Create production-ready inference class and demo script for end-to-end ECG classification.

### Tasks

- [ ] Create `ECGClassifier` production class
- [ ] Implement end-to-end prediction pipeline
- [ ] Add input validation and error handling
- [ ] Return predictions with confidence scores
- [ ] Create interactive demo script
- [ ] Add sample ECG files for testing
- [ ] Write usage documentation
- [ ] (Optional) Export model to ONNX

### Files to Create/Modify

| File | Action | Description |
|------|--------|-------------|
| `src/inference.py` | Create | Production inference class |
| `scripts/demo.py` | Create | Interactive demo |
| `examples/` | Create | Sample ECG files |

### Code Reference

```python
import torch
import numpy as np
from typing import List, Dict, Optional

class ECGClassifier:
    """Production-ready ECG arrhythmia classifier."""

    CLASS_NAMES = ['Normal', 'Supraventricular', 'Ventricular',
                   'Fusion', 'Unknown']

    def __init__(self, model_path: str, model_type: str = 'cnn',
                 device: str = 'auto'):
        if device == 'auto':
            self.device = 'cuda' if torch.cuda.is_available() else 'cpu'
        else:
            self.device = device

        self.model = self._load_model(model_path, model_type)
        self.preprocessor = ECGPreprocessor()

    def _load_model(self, path: str, model_type: str):
        if model_type == 'cnn':
            model = CNN1D(num_classes=5)
        elif model_type == 'lstm':
            model = LSTMAttention(num_classes=5)
        else:
            raise ValueError(f"Unknown model type: {model_type}")

        model.load_state_dict(torch.load(path, map_location=self.device))
        model.to(self.device)
        model.eval()
        return model

    def predict(self, ecg_signal: np.ndarray,
                fs: int = 360) -> List[Dict]:
        """Classify arrhythmias in ECG signal.

        Args:
            ecg_signal: Raw ECG signal (1D array)
            fs: Sampling frequency (default 360 Hz)

        Returns:
            List of dicts with beat predictions and confidence
        """
        # Validate input
        if not isinstance(ecg_signal, np.ndarray):
            ecg_signal = np.array(ecg_signal)
        if ecg_signal.ndim != 1:
            raise ValueError("ECG signal must be 1D array")
        if len(ecg_signal) < 720:
            raise ValueError("Signal too short (need at least 2 seconds)")

        # Preprocess
        filtered = self.preprocessor.bandpass_filter(ecg_signal)
        filtered = self.preprocessor.remove_baseline_wander(filtered)
        r_peaks = self.preprocessor.detect_r_peaks(filtered)

        if len(r_peaks) == 0:
            return []

        beats = self.preprocessor.segment_beats(filtered, r_peaks)

        # Predict
        results = []
        with torch.no_grad():
            for i, beat in enumerate(beats):
                tensor = torch.FloatTensor(beat).unsqueeze(0).unsqueeze(0)
                tensor = tensor.to(self.device)

                logits = self.model(tensor)
                probs = torch.softmax(logits, dim=1)
                pred = probs.argmax(dim=1).item()
                confidence = probs[0, pred].item()

                results.append({
                    'beat_index': i,
                    'r_peak_sample': int(r_peaks[i]),
                    'r_peak_time': float(r_peaks[i] / fs),
                    'prediction': self.CLASS_NAMES[pred],
                    'confidence': float(confidence),
                    'all_probabilities': {
                        name: float(probs[0, j])
                        for j, name in enumerate(self.CLASS_NAMES)
                    }
                })

        return results

    def predict_file(self, file_path: str) -> List[Dict]:
        """Load and classify ECG from file."""
        import wfdb
        record = wfdb.rdrecord(file_path)
        signal = record.p_signal[:, 0]  # First channel
        return self.predict(signal, fs=record.fs)
```

### Demo Script

```python
#!/usr/bin/env python3
"""Interactive demo for ECG classification."""

import argparse
from inference import ECGClassifier

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('input', help='ECG file path')
    parser.add_argument('--model', default='models/cnn_best.pt')
    args = parser.parse_args()

    classifier = ECGClassifier(args.model)
    results = classifier.predict_file(args.input)

    print(f"\nAnalyzed {len(results)} beats:\n")
    for r in results:
        print(f"Beat {r['beat_index']} @ {r['r_peak_time']:.2f}s: "
              f"{r['prediction']} ({r['confidence']:.1%})")

    # Summary
    from collections import Counter
    counts = Counter(r['prediction'] for r in results)
    print(f"\nSummary: {dict(counts)}")

if __name__ == '__main__':
    main()
```

### Definition of Done

- [ ] `ECGClassifier` class loads model and runs inference
- [ ] Input validation catches common errors
- [ ] Predictions include confidence scores
- [ ] Demo script works with MIT-BIH records
- [ ] Error messages are helpful for debugging
- [ ] Documentation explains usage

### Technical Notes

**For junior developers:**
- Always use `model.eval()` and `torch.no_grad()` for inference
- Validate inputs before processing (fail fast)
- Return confidence scores for clinical decision support
- Consider ONNX export for deployment without PyTorch

**Comment by nicolapierre** [comment]
Certainly! It seems like there might have been an error with the issue description since "[object Object]" was displayed. However, I'll provide a general response for handling production inference. If there are specific details or questions you intended to include, please feel free to add them.

---

Hello @username,

Thank you for reaching out regarding the production inference phase. Here are general steps and considerations to ensure a successful deployment of your inference model to production:

1. **Model Optimization**:
   - Ensure your model is optimized for inference. Consider techniques such as quantization, pruning, or using optimized libraries like TensorRT (for NVIDIA GPUs) or ONNX Runtime.

2. **Environment Setup**:
   - Confirm that your production environment matches the specifications used during development. This includes Python version, library dependencies, and compatible hardware (CPU/GPU).

3. **Inference Pipeline**:
   - Design a robust inference pipeline that can handle incoming requests efficiently. Consider using a framework like TensorFlow Serving, TorchServe, or FastAPI for deploying your model as a service.

4. **Scalability**:
   - Implement auto-scaling solutions if your application requires handling varying loads. Kubernetes or serverless platforms can help manage scaling efficiently.

5. **Monitoring and Logging**:
   - Integrate monitoring and logging to track model performance, latency, and errors in real-time. Tools like Prometheus, Grafana, and ELK Stack can be useful.

6. **Security**:
   - Ensure your inference service is secure. Implement authentication and encryption for data in transit and at rest.

7. **Testing**:
   - Conduct thorough testing in a staging environment to ensure accuracy and performance meet expectations before going live.

8. **Documentation**:
   - Update documentation with details on the deployment process, configurations, and any specific instructions for troubleshooting.

If you have specific questions or encounter any issues during this phase, please provide more details, and I’d be glad to assist further.

Best regards,

[your name]
